pkg/policy/api/ingress.go

```diff
@@ -108,12 +108,35 @@ type IngressRule struct {
 	//
 	// +optional
 	FromEntities EntitySlice `json:"fromEntities,omitempty"`
+
+	aggregatedSelectors EndpointSelectorSlice
+}
+
+// SetAggregatedSelectors creates a single slice containing all of the following
+// fields within the IngressRule, converted to EndpointSelector, to be stored
+// within the IngressRule for easy lookup while performing policy evaluation
+// for the rule:
+// * FromEntities
+// * FromCIDR
+// * FromCIDRSet
+//
+// FromEndpoints is not aggregated due to requirement folding in
+// GetSourceEndpointSelectorsWithRequirements()
+func (i *IngressRule) SetAggregatedSelectors() {
+	res := make(EndpointSelectorSlice, 0, len(i.FromEndpoints)+len(i.FromEntities)+len(i.FromCIDRSet)+len(i.FromCIDR))
+	res = append(res, i.FromEntities.GetAsEndpointSelectors()...)
+	res = append(res, i.FromCIDR.GetAsEndpointSelectors()...)
+	res = append(res, i.FromCIDRSet.GetAsEndpointSelectors()...)
+	i.aggregatedSelectors = res
 }
 
 // GetSourceEndpointSelectorsWithRequirements returns a slice of endpoints selectors covering
 // all L3 source selectors of the ingress rule
 func (i *IngressRule) GetSourceEndpointSelectorsWithRequirements(requirements []metav1.LabelSelectorRequirement) EndpointSelectorSlice {
-	res := make(EndpointSelectorSlice, 0, len(i.FromEndpoints)+len(i.FromEntities)+len(i.FromCIDRSet)+len(i.FromCIDR))
+	if i.aggregatedSelectors == nil {
+		i.SetAggregatedSelectors()
```

Comment:
There is no protection against parallel usage of `IngressRule` for this to happen safely.

Reply:
It's safe even if multiple goroutines run this in parallel. They will compute the same result so the order in which they set the pointer does not matter. I've added a comment to this effect.

Reply:
this feels really odd... also, @jrajahalme comment not added :-)

Reply:
The comment is in `SetAggregateSelectors()`, where the pointer is being set :-)